### DISCUSSION TITLE
Prevent automatic playback?

### DISCUSSION THREAD

**Discussion by macca007x** [spessasus/SpessaSynth#62] [Q&A] [answered]
Hey, how do I prevent the sequencer/synth from automatically playing as soon as the midifile loads?

It seems to be on by default.

Even if I add a `.pause()` directly after, it still plays for a brief moment??

```        
const sequencer = new Sequencer([{ binary: midiFileArrayBuffer }], synth);
sequencer.pause();
```

I want the user to load the midi, then press play when they are ready.

**Comment by macca007x** [comment]
@spessasus Could you provide some insight here? Not sure if I'm configuring incorrectly or not...

**Reply by macca007x** [reply]
I've managed to come up with a workaround, which involves manually setting the volume and playback.

This feels messy though. It would be good to have an `autoPlayback` toggle in the sequencer options.

```
        const newSynth = new Synthetizer(
          context.destination,
          soundFontArrayBuffer
        );

        newSynth.setMainVolume(0); // Set volume to 0 to avoid hearing the initial burst of sound

        const newSequencer = new Sequencer(
          [{ binary: midiFileArrayBuffer }],
          newSynth
        );

        newSequencer.pause(); // Stop the automated playback

        newSequencer.currentTime = 0; // Reset the position to 0, as the playback progresses slightly before .pause() is run

        newSynth.setMainVolume(masterVolume); // Reset volume so we can hear
```

**Reply by spessasus** [reply]
Yes, the sequencer currently does not have an option to disable auto start. I'll add that when I have the time.

**Reply by spessasus** [reply]
Hi, I've added this option to the lib (check the wiki page for the sequencer). Update your package (make sure to update worklet_processor.min.js too!)

**Reply by macca007x** [reply]
@spessasus Legend! Thank you :D

**Reply by macca007x** [reply]
@spessasus Hey I'm not sure the changes are working...

I even wiped and reinstalled the npm package (the files & worker are definitely up-to-date)

This has no effect, the song still plays on-load:

```
        const sequencer = new Sequencer(
          [{ binary: midiFileArrayBuffer }],
          synth,
          { skipToFirstNoteOn: false, autoPlay: false }
        );
```

**Reply by spessasus** [reply]
Hi, 
They work for me. This is the example I used:
```html
    <input type="file" id="mid">
    <input type="file" id="sf">
    <button id="go">Start</button>
    <script src="dist/out.bundle.js" type="module"></script>
```
```js
import { Synthetizer, Sequencer } from "spessasynth_lib"

document.getElementById("go").onclick = async () => {
    const context = new AudioContext();
    const sf = document.getElementById("sf").files[0];
    await context.audioWorklet.addModule(new URL("worklet_processor.min.js", import.meta.url));
    const synth = new Synthetizer(context.destination, await sf.arrayBuffer());
    const mid = await document.getElementById("mid").files[0].arrayBuffer();
    const seq = new Sequencer([{binary: mid}], synth, {
        autoPlay: false
    });

    // wait 5 seconds before starting playback
    await new Promise(r => setTimeout(r, 5000));
    seq.play();
}
```
The playback correctly starts after 5 seconds.

Make sure that you have:
1. replaced the old `worklet_processor.min.js` with the new one
2. Re-run the bundler you use
3. hard-reloaded the browser via `ctrl + shift + R`. Browsers tend to cache the old worklet files which result in missing the old behavior

I hope this helps!

PS: To get that syntax highlighting in markdown, type `js` at the first line, like this

```
```js
``````

**Reply by macca007x** [reply]
Hey! @spessasus Thanks for the detailed response, I appreciate your patience here.

So I've found the root cause of my issue: 

Seeking, and adjusting playback speed _also_ cause the sequencer to play (regardless of whether it is currently paused).

I had 2 useEffect hooks in the file, which listen for user seek/tempo adjustments.

I need this to run on initialisation, as the tempo adjustment should persist when changing songs:

```js
  useEffect(() => {
    if (sequencer) {
      sequencer.playbackRate = playbackSpeed;
    }
  }, [playbackSpeed, sequencer]);

  useEffect(() => {
    if (sequencer) {
      sequencer.currentTime = seekEvent.time / 1000;
    }
  }, [seekEvent]);
```

Looks to be from this:
![image](https://github.com/user-attachments/assets/78281858-ba58-4511-b9f8-76d7f0bd3746)

**Reply by macca007x** [reply]
This seems to be unexpected?

EG: When watching YouTube, I wouldn't expect seeking or changing video speed to make the video resume playback... only the play/pause button should does this.

**Reply by macca007x** [reply]
(It should honor the playback state present before the seek/playback event)

**Reply by spessasus** [reply]
> This seems to be unexpected?
> 
> EG: When watching YouTube, I wouldn't expect seeking or changing video speed to make the video resume playback... only the play/pause button should does this.

This is the intended behavior, it was implemented that way from the beginning. That's because if we seek, we effectively *play to* that position. We can't just *go* to that time. When you pause, you don't want any actions to be sent to the synthesizer. And when you seek, we have to prepare the synth by playing all cc and the program changes up to that point (and NRPNs and sysExes) to correctly resume the playback. For example, if MIDI selects trumpet on channel 1, then square wave, we have to check which one should be set at that time.

I could change it so setting the current time does not start the playback, but that would probably break a lot of implementations... I'm not sure. Because for now you can pause, let the user seek, and when they hit play then you set `currentTime` to achieve the desired effect.

**Reply by macca007x** [reply]
I see... sorry to hear it's a complex change.

I'll have to try and rethink the front-end implementation.

**Reply by spessasus** [reply]
> I see... sorry to hear it's a complex change.
> 
> I'll have to try and rethink the front-end implementation.

No, it's rather simple to implement, but I'm not sure if it will break other people's programs that assume the current behavior.

**Reply by macca007x** [reply]
Oh really? What would I need to change? I can make a fork

**Reply by spessasus** [reply]
I've created the [discussion62](https://github.com/spessasus/SpessaSynth/tree/discussion62) branch with the requested functionality. Feel free to test on it.

**Reply by macca007x** [reply]
@spessasus That works perfectly, thanks man!

**Reply by spessasus** [reply]
I think I have an idea on how to integrate this with main.

Sequencer would have a property `preservePlaybackState` which would default to false, indicating the current behavior. Once you set it to true, it will work like the branch I made for you.

What do you think?

**Reply by macca007x** [reply]
@spessasus Sounds great 👌

**Reply by spessasus** [reply, marked as answer]
Hi again,

I've added that functionality into the sequencer. Upgrade to the latest npm package and check out https://github.com/spessasus/SpessaSynth/wiki/Sequencer-Class

I hope this was what you were looking for. Feel free to reply if something isn't right.
